### PR TITLE
Refactor hardware extension API to refer to "target" instead.

### PR DESCRIPTION
**Pull request by stuartarchibald**
As title. This is a largely mechanical refactor. It changes the
new hardware extension API to refer to "target" instead of
"hardware". This is because the "hardware" is often actually
synthetic and "target" better describes this.

### CODE REVIEW

**Comment by stuartarchibald** [comment]
Notes:
1. This will need a CUDA smoke test because the CUDA target makes use of this API for `@overload` support.
2. This PR probably blocks any further refactoring in the `@overload` area as it renames most of the API in use (it could still be done, the merge conflicts might not be easy to deal with though).

**Review by gmarkall — COMMENTED** [review]
This looks good and tests OK with the CUDA target for me locally on Linux.

I've looked through the CUDA changes and they look fine - I've also read through the rest of the PR and it seems OK to the extent that I could understand the changes.

I've made one small suggestion on the diff, but it's not a deal breaker for me. :-)

File: numba/cuda/initialize.py [hunk 1]
```diff
@@ -4,14 +4,14 @@ def initialize_all():
...
+                                             dispatcher_registry,
+                                             jit_registry)
 
     def cuda_jit_device(*args, **kwargs):
         kwargs['device'] = True
         return cuda.jit(*args, **kwargs)
 
-    cuda_hw = hardware_registry["cuda"]
-    decorators.jit_registry[cuda_hw] = cuda_jit_device
+    cuda_hw = target_registry["cuda"]
```
[the remark this answers is not in the document]
Comment (stuartarchibald) [reply numba/cuda/initialize.py]: Thanks, agree, patched in 2c85862

**Review by gmarkall — APPROVED** [review]
LGTM!

**Comment by sklam** [comment]
BFID `numba_smoketest_cuda_yaml_57`

**Pull request merged**